File: scripts/compose-poster.mjs
/**
 * Overlay a real scannable QR (with logo) onto the designed invite poster.
 * Uses Jimp so it works on Node 18.
 */
import QRCode from 'qrcode'
import { Jimp } from 'jimp'
import path from 'node:path'
import { fileURLToPath } from 'node:url'

const __dirname = path.dirname(fileURLToPath(import.meta.url))
const root = path.resolve(__dirname, '..')
const publicDir = path.join(root, 'public')

const REGISTER_URL = 'https://myummah.co.uk/#connect'
const posterPath = path.join(publicDir, 'mosque-invite-poster.png')
const logoPath = path.join(publicDir, 'logo.png')
const qrOutPath = path.join(publicDir, 'qr-register.png')
const finalOutPath = path.join(publicDir, 'mosque-invite-final.png')

const poster = await Jimp.read(posterPath)
const { width, height } = poster.bitmap

// QR sits in the lower-center of the designed poster (approx. from visual layout)
const qrSize = Math.round(width * 0.36)
const qrX = Math.round((width - qrSize) / 2)
const qrY = Math.round(height * 0.545)

const qrBuffer = await QRCode.toBuffer(REGISTER_URL, {
  width: qrSize,
  margin: 1,
  errorCorrectionLevel: 'H',
  color: { dark: '#0a2240ff', light: '#ffffffff' },
})

const qr = await Jimp.read(qrBuffer)

// Center logo badge on QR (needs high error correction)
const logo = await Jimp.read(logoPath)
const badge = Math.round(qrSize * 0.22)
logo.cover({ w: badge, h: badge })

const pad = Math.round(badge * 0.16)
const badgeBg = new Jimp({
  width: badge + pad * 2,
  height: badge + pad * 2,
  color: 0xffffffff,
})
badgeBg.composite(logo, pad, pad)
qr.composite(
  badgeBg,
  Math.round((qrSize - badgeBg.bitmap.width) / 2),
  Math.round((qrSize - badgeBg.bitmap.height) / 2),
)

await qr.write(qrOutPath)

// Soft white plate behind QR to cover the old non-scannable code cleanly
const platePad = Math.round(qrSize * 0.06)
const plate = new Jimp({
  width: qrSize + platePad * 2,
  height: qrSize + platePad * 2,
  color: 0xffffffff,
})
poster.composite(plate, qrX - platePad, qrY - platePad)
poster.composite(qr, qrX, qrY)

await poster.write(finalOutPath)

console.log(`Poster ${width}x${height}`)
console.log(`QR placed at (${qrX}, ${qrY}) size ${qrSize}`)
console.log(`Wrote ${qrOutPath}`)
console.log(`Wrote ${finalOutPath}`)

File: scripts/compose-poster.py
"""Overlay a real scannable QR (with logo) onto the designed invite poster."""
from __future__ import annotations

from pathlib import Path

import qrcode
from PIL import Image, ImageDraw

ROOT = Path(__file__).resolve().parents[1]
PUBLIC = ROOT / "public"
REGISTER_URL = "https://myummah.co.uk/#connect"

poster_path = PUBLIC / "mosque-invite-poster.png"
logo_path = PUBLIC / "logo.png"
qr_out = PUBLIC / "qr-register.png"
final_out = PUBLIC / "mosque-invite-final.png"


def main() -> None:
    poster = Image.open(poster_path).convert("RGBA")
    width, height = poster.size

    # Lower-center QR area on the designed poster
    qr_size = int(width * 0.36)
    qr_x = (width - qr_size) // 2
    qr_y = int(height * 0.545)

    qr = qrcode.QRCode(
        version=None,
        error_correction=qrcode.constants.ERROR_CORRECT_H,
        box_size=12,
        border=2,
    )
    qr.add_data(REGISTER_URL)
    qr.make(fit=True)
    qr_img = qr.make_image(fill_color="#0a2240", back_color="white").convert("RGBA")
    qr_img = qr_img.resize((qr_size, qr_size), Image.Resampling.NEAREST)

    # Logo badge in QR center
    logo = Image.open(logo_path).convert("RGBA")
    badge = int(qr_size * 0.22)
    logo = logo.resize((badge, badge), Image.Resampling.LANCZOS)
    pad = max(4, int(badge * 0.16))
    badge_bg = Image.new("RGBA", (badge + pad * 2, badge + pad * 2), (255, 255, 255, 255))
    # Rounded-ish plate via simple paste; keep crisp for scan reliability
    badge_bg.paste(logo, (pad, pad), logo)
    bx = (qr_size - badge_bg.width) // 2
    by = (qr_size - badge_bg.height) // 2
    qr_img.paste(badge_bg, (bx, by), badge_bg)

    qr_img.convert("RGB").save(qr_out, "PNG")

    # White plate to cover previous decorative QR
    plate_pad = int(qr_size * 0.07)
    plate = Image.new(
        "RGBA",
        (qr_size + plate_pad * 2, qr_size + plate_pad * 2),
        (255, 255, 255, 255),
    )
    # Soft gold border
    draw = ImageDraw.Draw(plate)
    draw.rounded_rectangle(
        [0, 0, plate.width - 1, plate.height - 1],
        radius=max(12, plate_pad),
        outline=(212, 175, 55, 255),
        width=max(3, plate_pad // 4),
    )

    poster.paste(plate, (qr_x - plate_pad, qr_y - plate_pad), plate)
    poster.paste(qr_img, (qr_x, qr_y), qr_img)

    poster.convert("RGB").save(final_out, "PNG", quality=95)
    print(f"Poster {width}x{height}")
    print(f"QR at ({qr_x}, {qr_y}) size {qr_size}")
    print(f"Wrote {qr_out}")
    print(f"Wrote {final_out}")


if __name__ == "__main__":
    main()
